src/NativeBridge/PythonInterop.cpp

```diff
@@ -1,55 +1,266 @@
+        dict[name] = np::from_data(
+            data,
+            np::dtype::get_builtin<T>(),
+            bp::make_tuple(_pData->size()),
+            bp::make_tuple(sizeof(T)), bp::object(h));
+    }
+    break;
     }
+}
 
-    return found->second(kind, numRows, numCols);
+template <>
+void PythonObjectSingle<std::string>::AddToDict(bp::dict& dict,
+                                                const std::string& name,
+                                                const std::vector<std::string>* keyNames,
+                                                const size_t expectedRows)
+{
+    bp::list list;
+    for (size_t i = 0; i < _pData->size(); i++)
+    {
+        bp::object obj;
+        const std::string& value = _pData->at(i);
+        if (!value.empty())
+        {
+            obj = bp::object(value);
+        }
+        list.append(obj);
+    }
+    dict[name] = list;
 }
 
-template <class T> PythonObjectBase* PythonObjectBase::CreateObject(const int& kind, size_t nRows, size_t nColumns)
+/*
+ * Note: an instance of this object should not be used
+ * and should be considered invalid after the first time
+ * this method has been called.
+ */
+template <class T, class T2>
+void PythonObjectVariable<T, T2>::Deleter(PyObject* obj)
 {
-    return new PythonObject<T>(kind, nRows, nColumns);
+    auto* deleteData = static_cast<PythonObjectVariable<T, T2>::DeleteData*>(PyCapsule_GetPointer(obj, NULL));
+
+    PythonObjectVariable<T, T2>* instance = deleteData->instance;
+    size_t column = deleteData->column;
+
+    std::vector<T2>* data = instance->_data[column];
+    if (data != nullptr)
+    {
+        instance->_data[column] = nullptr;
+        instance->_numDeletedColumns++;
+        delete data;
+
+        if (instance->_numDeletedColumns == instance->_data.size())
+        {
+            delete instance;
+        }
+    }
 }
 
+template<class T, class T2>
+inline void PythonObjectVariable<T, T2>::AddToDict(bp::dict& dict,
+                                                   const std::string& name,
+                                                   const std::vector<std::string>* keyNames,
+                                                   const size_t expectedRows)
+{
+    size_t numRows = (expectedRows > this->_numRows) ? expectedRows : this->_numRows;
+    size_t numCols = _data.size();
+
+    if (numCols == 0)
+    {
+        /*
+         * If there were no values set then create
+         * a column so it can be filled with NANs.
+         */
+        _data.push_back(new std::vector<T2>());
+        numCols = 1;
+    }
+
+    const std::string colNameBase = name + ".";
+    int maxDigits = (int)ceil(log10(numCols));
+    if (maxDigits == 0) maxDigits = 1;
+
+    for (size_t i = 0; i < numCols; i++)
+    {
+        /*
+         * Make sure all the columns are the same length.
+         */
+        for (size_t j = _data[i]->size(); j < numRows; j++)
+        {
+            _data[i]->push_back(NAN);
```

Comment:
`_data[i]` can be called multiple times here even though its always the same value. Could this be sped up by creating a local variable instead.

Reply:
Fixed in commit e97079e796a9b2c11eb327bf8febb46fcff74bf9